Pytorch example with DataLoader adapter, using MNIST data

forbearer opened the pull request:
This code includes an MNIST dataset generator, a pytorch training example that uses the resulting dataset, and a simple `README.md`.

As can be seen from the `main.py`, there are few _limitations_ that come to light which could help us improve petastorm:
* Batch shuffling
* Support for custom transforms
* Total data size (or some semblance of it?)

Running `pytorch/examples/mnist/main.py` (in a Docker container) with the default 10 epoch yielded the following outcome (I just show the test output for the middle 8 epochs):
```
...
Train Epoch: 1 [59520/60000 (99%)]	Loss: 0.505042

Test set: Average loss: 0.2056, Accuracy: 9395/10000 (94%)

...
Test set: Average loss: 0.1337, Accuracy: 9596/10000 (96%)
Test set: Average loss: 0.1033, Accuracy: 9684/10000 (97%)
Test set: Average loss: 0.0919, Accuracy: 9710/10000 (97%)
Test set: Average loss: 0.0760, Accuracy: 9770/10000 (98%)
Test set: Average loss: 0.0689, Accuracy: 9797/10000 (98%)
Test set: Average loss: 0.0623, Accuracy: 9803/10000 (98%)
Test set: Average loss: 0.0632, Accuracy: 9791/10000 (98%)
Test set: Average loss: 0.0541, Accuracy: 9818/10000 (98%)

...
Train Epoch: 10 [59520/60000 (99%)]	Loss: 0.040862

Test set: Average loss: 0.0505, Accuracy: 9845/10000 (98%)

real	3m3.021s
user	20m4.680s
sys	0m22.228s
```
With the petastormed variant, the training accuracy looks on-par, with somewhat better runtime. I'll show just the test output:
```
Test set: Average loss: 0.2035, Accuracy: 9385/10000 (94%)
Test set: Average loss: 0.1326, Accuracy: 9591/10000 (96%)
Test set: Average loss: 0.1040, Accuracy: 9675/10000 (97%)
Test set: Average loss: 0.0887, Accuracy: 9705/10000 (97%)
Test set: Average loss: 0.0761, Accuracy: 9752/10000 (98%)
Test set: Average loss: 0.0715, Accuracy: 9774/10000 (98%)
Test set: Average loss: 0.0627, Accuracy: 9797/10000 (98%)
Test set: Average loss: 0.0606, Accuracy: 9810/10000 (98%)
Test set: Average loss: 0.0582, Accuracy: 9824/10000 (98%)
Test set: Average loss: 0.0548, Accuracy: 9828/10000 (98%)

real	2m35.852s
user	2m33.508s
sys	0m6.576s
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/petastorm?pullRequest=50) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/petastorm?pullRequest=50) before we can accept your contribution.<br/><hr/>**Owen Cheng** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/petastorm?pullRequest=50) it.</sub>

forbearer commented:
@selitvin About DataLoader and num_epochs, you are right that it's not in the constructor.  I had `num_workers` in my memory and that got mixed up when I was talking to you.  My bad.  This does simplify things.  :-)

forbearer commented on examples/mnist/tests/test_generate_mnist_dataset.py:
```diff
@@ -0,0 +1,86 @@
+
+# This test requires torch import, but that's causing dlopen to fail in Travis run. :(
+# So skipping it for now
+def _dont_test_mnist_download(tmpdir):
```
https://github.com/uber/petastorm/blob/c59f0c8deefe2b1c6cbdf00b08260c22eadef165/examples/mnist/tests/test_generate_mnist_dataset.py#L72-L79

I've disabled this test for now due to issue dlopen'ing `torch` module in travis CI run.

If it's OK with you, I'd like to tackle that issue in a separate PR...as this one is getting quite long.

rgruener replied:
You should skip this test using `@pytest.mark.skip` instead of changing the function name

forbearer commented:
Aww yeah! I'd be happy to land this and we can improve on the test in a future PR?

selitvin commented on petastorm/pytorch.py:
```diff
@@ -0,0 +1,118 @@
+        """
+        batch = []
+        for row in self.reader:
+            batch.append(self.transform(row))
```
This transform can be heavy. We need to think about pushing this down to a worker. Will be easier with the new reader implementation. Not for this PR.

forbearer replied:
:-) Agreed!

selitvin commented on petastorm/pytorch.py:
```diff
@@ -0,0 +1,118 @@
+            if len(batch) == self.batch_size:
+                yield self.collate_fn(batch)
+                batch = []
+        if batch:
```
pytorch's `DataLoader` has `drop_last` to define whether we want an incomplete batch. Should be simple to add the implementation here. Ok if you want as a separate CL.

forbearer replied:
I think that makes sense. Let me add that in a separate CL/PR.

forbearer commented on .travis.yml:
```diff
@@ -20,9 +20,11 @@ python:
 
 install:
+  # Upgrade pip to avoid weird failures, such as failing to install torchvision
+  - pip install --upgrade pip
```
Sure. I think that makes sense. How about 18.0? :-)

rgruener replied:
Once were doing that it might make sense to install numpy 1.14.5 (though maybe this PR is large enough)

forbearer commented on petastorm/pytorch.py:
```diff
@@ -0,0 +1,144 @@
+                 training_partition=None, num_training_partitions=None,
+                 read_timeout_s=None, cache=None):
+        """
+        Initializes a batch reader object, with a default collate and optional transform functions.
```
oops. thx!

rgruener commented on README.md:
```diff
@@ -101,6 +107,43 @@ TODO: Reading a dataset from pytorch
+torch.manual_seed(1)
+device = torch.device('cpu')
+model = Net().to(device)
+optimizer = torch.optim.SGD(model.parameters(), lr=0,01, momentum=0.5)
```
is `lr=0,01` a typo?

forbearer replied:
Yeah, good eye! I just posted a fix. Thanks.

rgruener commented on petastorm/pytorch.py:
```diff
@@ -0,0 +1,118 @@
+    once more at the very end.
+    """
+
+    def __init__(self, dataset_url, schema_fields=None, batch_size=1, shuffle_options=None,
```
This class copies a lot of logic and boilerplate from the Reader class. This will require keeping a lot of things in sync, e.g. arguments, etc. Is there a reason we dont pass in an instantiated reader object to the DataLoader or have it be a parent class?

rgruener replied:
This would also make it more inline with what we do with tensorflow (have it be an adapter on top of the python reader instead of a different class)

forbearer replied:
Hmm, that's a good point.  I also didn't like reproducing reader params all over again.

Also, I did battle these two options out in my mind...didn't think to look at tf for an existing pattern!

OK, let me know reformulate this to pass in an instantiated reader.

rgruener commented on petastorm/reader.py:
```diff
@@ -284,6 +284,13 @@ def fetch(self, timeout=None):
         logger.warn(warning_message)
         return self._workers_pool.get_results(timeout=timeout)
 
+    def __len__(self):
```
please add a unit test for this function

forbearer replied:
Definitely. I added a new case to `test_end_to_end.py`.

Arguably, I should include a case to test raising `NotImplementedError`!

selitvin commented on examples/mnist/tests/test_generate_mnist_dataset.py:
```diff
@@ -0,0 +1,132 @@
+def test_image_to_numpy(small_mock_mnist_data):
+    """ Show output of image object reshaped as numpy array """
+    im = small_mock_mnist_data['train'][0]
+    print(im)
```
Should not have raw prints in the test.

forbearer replied:
Ah, agreed.  This apparent slows down test run!

I'll use logging to allow getting output with `pytest -s`.

selitvin commented on examples/mnist/tests/test_generate_mnist_dataset.py:
```diff
@@ -0,0 +1,132 @@
+    im = small_mock_mnist_data['train'][0]
+    print(im)
+    print(im[1])
+    assert im[1] >= 0 and im[1] <= 9
```
Nitpick: `0<=im[1]<=9`

forbearer replied:
Yes. Pythonic!

selitvin commented on examples/mnist/tests/test_generate_mnist_dataset.py:
```diff
@@ -0,0 +1,132 @@
+
+# This test requires torch import, but that's causing dlopen to fail in Travis run. :(
+# So skipping it for now
+def _dont_test_mnist_download(tmpdir):
```
Consider using a standard unittest feature:
```
@unittest.skip('some description')
```

selitvin commented on petastorm/pytorch.py:
```diff
@@ -0,0 +1,118 @@
+                 training_partition=None, num_training_partitions=None,
+                 read_timeout_s=None, cache=None):
+        """
+        Initializes a data loader object, with a default collate and optional transform functions.
```
A question: would it make sense to pass a reader instance as a parameter to the DatasetLoader instead of forwarding all arguments to it? Kinda same as pytorch's Loader takes an instance of a `Dataset`?

And on the same note, the `reader_pool` is not forwarded, which emphasizes the point :)

forbearer replied:
Agreed! @rgruener also commented on this exact same thing. Great minds! :-)

Posting an updated commit real soon that addresses all these review comments.  Thank you so much!

forbearer commented:
@selitvin @rgruener Thank you both for your detailed round-2 reviews.

I've addressed all the comments in the new commit, re-amended and pushed.  I'll be checking on the test to get them passing, given two issues I've encountered so far:
1) dlopen error when importing `torch`
2) seg fault when running either `test_generate` or `test_full_pytorch_example`

forbearer commented:
OK, the good news is that the CI passes with Python 3.x!

forbearer commented:
Lacking access to travis debug mode, I'm trying various `pytest` options to ferret out the issue.

One thought is that `pytest` execution order is presumably deterministic, and the collect is always failing on item 129, with 1 failing, which means the 130th test is bad (or maybe 131st).

These are test cases 129, 130, 131 when the full list of tests are enumerated:
```
petastorm/tests/test_predicates.py::PredicatesTest::test_inclusion PASSED                                                                                                                                       [ 65%]
petastorm/tests/test_predicates.py::PredicatesTest::test_list_inclusion PASSED                                                                                                                                  [ 66%]
petastorm/tests/test_predicates.py::PredicatesTest::test_negation PASSED                                                                                                                                        [ 66%]
```

forbearer commented:
Removing the various pytest runs as they weren't yielding additional information.

FYI, this generates a list of test functions, without running them:
```
pytest --collect-only | grep Function
```

I can't see why any of the first 3 test cases in `test_predicates.py` would cause seg fault...definitely no pytorch import either :-)

forbearer commented:
Filtering out `test_predicates.py` didn't help the pytest collect seg-fault.  That's not surprising.

I decided to try out filtering out `test_pytorch_utils.py`.

forbearer commented:
Well, ladies and gentlemen, `test_pytorch_utils.py` was the culprit of the segfault during pytest collect.

Next: figuring out why...

forbearer commented:
Wow, so based on this (https://github.com/pytorch/pytorch/issues/2083) and some experimentation in the travis docker container, I was able to resolve the 2.7 torch import issue (with dlopen running out of static TLS) by importing torch *way early* in conftest.py.

That's really awful, but solves all of the problem surrounding my unit tests needing to be skipped.

I'll post that change as part of this PR.  Let me know what you think.

forbearer commented:
Strangely, this test is reliably seg-faulting now, even when I do `pytest -v` locally:
```
petastorm/tests/test_end_to_end.py::test_simple_read[reader_factory1] /home/travis/.travis/job_stages: line 78:  2933 Segmentation fault      (core dumped) pytest -v --cov=./
```

forbearer commented:
The other lingering issue is captured in #52.  For this, I will skip the mnist generate and pytorch example test cases for Python 2.x.

forbearer commented:
Well, the experiment to upgrade to `xenial` and `bionic` did not yield positive result.

So I started wondering if the dlopen static TLS problem was actually incorporated into those two Ubuntu releases.

From this https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1673956, the bug was analyzed and bugfixed in https://sourceware.org/bugzilla/show_bug.cgi?id=17620

According to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=793641, the above fix is in `glibc-2.22`.

According to ubuntu, xenial http://changelogs.ubuntu.com/changelogs/pool/main/g/glibc/glibc_2.23-0ubuntu10/changelog has `glibc-2.22`

Well, as I still encounter the same dlopen static TLS error in both xenial and bionic, why might that be?

forbearer commented:
LOL: https://travis-ci.com/uber/petastorm/jobs/141963848

I was so stoked that the build passed, but discovered upon closer examination that the problem persists: seg fault during collect.

At this point, here is what I have tried and what I know:
* So long as test_pytorch_utils.py is in the pytest path, collect fails, even if using `-k` to exclude
* Different versions of Ubuntu did not resolve the issue, tried: `trusty`, `xenial`, and `bionic`, despite the fact that `glibc-2.22` (which **is** in `xenial` onward) contains the static TLS fix (*by increasing DTV surplus slots from 14 to 32, or some slightly larger integer*)
* Taking Clearwater's advice, I tried installing libgomp1 and preceding pytest with `export LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libgomp.so.1"`, but that did not resolve either!
* Using these commands (adapted from https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=793689#24), I believe the culprint *may* be that the `torch-0.4.1` we are pip installing statically linked against a specific libgomp, and therefore override does not work!
  * If I forcibly symlink torch's libgomp, I get the GOMP_4.0 not found error
  * Build from source *should* work, according to https://github.com/pytorch/pytorch/issues/643#issuecomment-292744036
```
for lib in $(ldd $(find /usr -name _C.so |& grep -v denied) | grep "=>" | sed 's/.*=> \([/a-z0-9\._+-]*\) .*/\1/g' | sort); do echo "readelf -d -W $lib"; readelf -d -W "$lib"; done | egrep 'readelf|STATIC_TLS' | grep -B1 TLS

readelf -d -W /lib/x86_64-linux-gnu/libc.so.6
 0x000000000000001e (FLAGS)              STATIC_TLS
--
readelf -d -W /lib/x86_64-linux-gnu/libm.so.6
 0x000000000000001e (FLAGS)              STATIC_TLS
readelf -d -W /lib/x86_64-linux-gnu/libpthread.so.0
 0x000000000000001e (FLAGS)              STATIC_TLS
readelf -d -W /lib/x86_64-linux-gnu/librt.so.1
 0x000000000000001e (FLAGS)              STATIC_TLS
--
readelf -d -W /usr/local/lib/python2.7/dist-packages/torch/lib/libgomp-7bcb08ae.so.1
 0x000000000000001e (FLAGS)              STATIC_TLS
```

Below, I do LD_PRELOAD first to show that `_C.so` still pulls torch's libgomp:
```
$ export LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libgomp.so.1"; for lib in $(ldd $(find /usr -name _C.so |& grep -v denied) | grep "=>" | sed 's/.*=> \([/a-z0-9\._+-]*\) .*/\1/g' | sort); do echo "readelf -d -W $lib"; readelf -d -W "$lib"; done | egrep 'readelf|STATIC_TLS' | grep -B1 TLS; unset LD_PRELOAD
readelf: Error: 'libnvToolsExt-3965bdd0.so.1': No such file
readelf: Error: '=>': No such file
readelf: Error: '(0x00007ff8e5dd3000)': No such file
readelf -d -W /lib/x86_64-linux-gnu/libc.so.6
 0x000000000000001e (FLAGS)              STATIC_TLS
--
readelf -d -W /lib/x86_64-linux-gnu/libm.so.6
 0x000000000000001e (FLAGS)              STATIC_TLS
readelf -d -W /lib/x86_64-linux-gnu/libpthread.so.0
 0x000000000000001e (FLAGS)              STATIC_TLS
readelf -d -W /lib/x86_64-linux-gnu/librt.so.1
 0x000000000000001e (FLAGS)              STATIC_TLS
--
readelf -d -W /usr/local/lib/python2.7/dist-packages/torch/lib/libgomp-7bcb08ae.so.1
 0x000000000000001e (FLAGS)              STATIC_TLS
```

**Last ditch effort**: I'm going to try pip installing the [CPU torch](http://download.pytorch.org/whl/cpu/torch-0.4.1-cp27-cp27mu-linux_x86_64.whl), followed by building from source using 
[Dockerfile](https://github.com/pytorch/pytorch/blob/master/docker/pytorch/Dockerfile). If neither of these efforts work, then I'm going to split pytorch tests from the rest of the pytest in order to make progress.

This PR has sunk quite a bit of time.  :-\

forbearer commented on examples/mnist/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+FROM ubuntu:14.04.5
```
This Dockerfile is not meant to be committed (unless we want to for some reason).

I'm just trying to track what I've done.

forbearer commented on .travis.yml:
```diff
@@ -12,24 +12,29 @@
 # limitations under the License.
 
-dist: trusty
+dist: xenial
```
Yes, this is a deliberate change.  I think we should proceed with xenial to be ahead of the game.

forbearer commented on examples/mnist/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+FROM pytorch:latest
```
Ah, yeah, that's actually a good point.

If there will be a tf example using MNIST, perhaps I should simply make a pytorch subdirectory. Then that would make things much clearer?

forbearer replied:
On second thought, actually, mnist generation leverages pytorch as well, so I don't think I'll gain much to introduce a pytorch subdirectory.  I'll simply add a legal header and comment about functionality to this Dockerfile.

The README.md in this directory explains use of the docker file.

forbearer commented on examples/mnist/main.py:
```diff
@@ -0,0 +1,158 @@
+
+    # Instantiate each petastorm Reader with a single thread, shuffle enabled, and appropriate epoch setting
+    for epoch in range(1, loop_epochs + 1):
+        with DataLoader(Reader('{}/train'.format(args.dataset_url), reader_pool=ThreadPool(1),
```
Both arguments `reader_pool` and `shuffle_options` are specified in this code, or do you mean, give it a unique name and then supply that?  E.g.,
```
single_thread = ThreadPool(1)
shuffle_enabled = ShuffleOptions()
...
with DataLoader(Reader(..., reader_pool=single_thread, shuffle_options=shuffle_enabled, ...)?
```

forbearer replied:
OK, I think I see what you are getting at.  Removing these two arguments would work just fine.  :-)
